mdadm: add new init features; bug fixes

josephtingiris opened the pull request:
Thanks for your contribution to OpenWrt!

To help keep the codebase consistent and readable,
and to help people review your contribution,
we ask you to follow the rules you find in the wiki at this link
https://openwrt.org/submitting-patches

Please remove this message before posting the pull request.

josephtingiris committed:
do not build with defunct device & uuid; comment out device & uuid

josephtingiris committed:
check block info for active linux_raid_members & dynamically update uci mdadm config prior to starting mdadm @ S13

josephtingiris committed:
mdadm: Install /etc/init.d/mdadm_config

josephtingiris commented:
Thanks for your insight.  That would explain why I couldn't find any examples in the other init scripts of explicit uci sets.  I understand the point that uci commit is a big hammer.

My first foray into OpenWrt was this week.   I appreciate the goal.  I'm new to this project, though not to other Linux distributions.  

Modern versions of Debian, Fedora, RHEL, et. al. dynamically assemble meta devices based on their drive header and automatically assign an unused device name e.g. /dev/md0.  There's generally no need for a human's intervention.  They've also all incorporated systemd and by comparison, older sysvinit styled distributions are lacking.

Having 'auto stuff' is helpful.  That's what computers do best.  The advantage is that humans don't have to do it.  These are tasks computers can do better, faster, and much more reliably.

I think the idea of having a uci option, e.g. 'auto_config' is a good idea. It should be the default.  If someone wants to change it and revert back to manual configuration then that should take precedence.  

I'll modify this PR to support both methods, as well as re-comment the mdadmin.config to remove the defunct uuid and enable auto_config.  When auto_config is used then it wont use uci for dynamic state.

Thanks again.

cshoredaniel commented:
actually mdadm already auto-assembles all MD devices based on what's there (although if you want to extended that to raid on top of another dm I'm not sure...but that's really getting outside of the scope of a router distribution IMO).

josephtingiris committed:
mdadm: Update init logic for autoconfig

This adds autoconfig support to mdadm.  /etc/config/mdadm now recognizes a uci 'defaults' section with an 'autoconfig' option.  When set, mdadm init will automatically determine the meta device UUIDs & appropriate device name(s).

The new mdadm_autoconfig() function will dynamically update /var/etc/mdadm.conf, rather than using uci.  If mdadm.@defaults[0].autoconfig is false then the previous manual logic will be used.

START & STOP order were also modified to allot more space before fstab starts, i.e for a cryptsetup block layer, and to stop mdadm last, to prevent conflicts with mdadm attempting to go offline prior to umount running.

Signed-off-by: Joseph Tingiris <joseph.tingiris@gmail.com>

josephtingiris commented on package/utils/mdadm/files/mdadm.config:
```diff
@@ -4,9 +4,9 @@ config mdadm
 	# list devices /dev/sd*
 	# list devices partitions
 
-config array
```
That's what got me started.  I've used mdadm a lot on other distros.  My expectation was that it'd auto-assemble the array but it didn't.  I traced it back to the Openwrt config & how its init builds /var/etc/mdadm.  Needless to say it's broken ootb.

It's apparently been this way for 2+ years and is still that way in openwrt/openwrt master.   I reported that with more detail, here.

https://bugs.openwrt.org/index.php?do=details&task_id=2043&opened=1797&type%5B0%5D=&sev%5B0%5D=&due%5B0%5D=&cat%5B0%5D=&status%5B0%5D=open&percent%5B0%5D=&reported%5B0%5D=

Once I'd noticed that my UUIDs didn't match the config the package installed, I tested first by commenting them out with the expectation that it'd scan and auto-assemble.  Didn't work.  Maybe my bad.  I've got a better grasp of uci & procd now as a result.

Updating the UUID in uci did work.  That's when I noticed /etc/init.d/umount is K99.  mdadm is K98.  That's the second, or third problem.  Rebooting causes mdadm stop to fail and umount remounts everything read only.  It's messy.  This should all work seamlessly as it does in other distros.

Further testing shows that auto-assemble does work *if* it's already been assembled at least once.  At this point, I'm shooting for security and stability.  There are obvious, albeit minor, issues with the way mdadm, cryptsetup, lvm2, and umount are set to init on Openwrt 18+

FWIW;  I've got raw -> md -> cryptsetup -> lvm2 -> fs. all working seamlessly.  I want it to stay working reliably, without concern that a package upgrade is going to break it.

cshoredaniel replied:
Ah, I checked the history -- I originally just did https://github.com/openwrt/openwrt/commit/813efe57e434037fb58bd3e16ebd3a1cfd6ceb82#diff-020e9c6e02016d2840b3f057677eb0dd which didn't really have an uci config but @jow- add uci config in https://github.com/openwrt/openwrt/commit/813efe57e434037fb58bd3e16ebd3a1cfd6ceb82#diff-020e9c6e02016d2840b3f057677eb0dd but apparently no one noticed for a long time there was an issue ... I wasn't active with OpenWrt/LEDE at the time and I guess it's not (yet) a major use case.

cshoredaniel replied:
I look forward to the result!

josephtingiris commented on package/utils/mdadm/files/mdadm_config.init:
```diff
@@ -1,151 +0,0 @@
-#!/bin/sh /etc/rc.common
```
I'd previously pushed it and subsequently deleted it.  I removed that logic per your suggestion to not use uci and integrated an 'autoconfig' feature that piggyback's on the original logic.  If 'autoconfig' is false then it'll use the old logic.  If it's on then it uses the new function I've added.

josephtingiris commented:
> (although if you want to extended that to raid on top of another dm I'm not sure...but that's really getting outside of the scope of a router distribution IMO).

My Netgear R7800s have eSATA and USB 3 ports.  Lots of modern routers are media gateways too.  It's in Netgear's, ASUS's, and many other proprietary OSs.  I want my media on a raid.  I want it encrypted.  I want to know if a drive fails all's not lost and if my drives get stolen then they better know what a block device and uuid is and how to get my key files.  I want to sleep at night knowing all that and if there's a power outage everything comes back online without me thinking about it beyond this (and a few other) threads.  OpenWrt via Linux has that promise, with a little work.

josephtingiris committed:
mdadm: add autoconfig feature; init bug fixes

This adds autoconfig support to mdadm.  /etc/config/mdadm now
recognizes a uci 'defaults' section with an 'autoconfig' option.
When set, mdadm init will automatically determine the meta device
UUIDs & appropriate device name(s).

The new mdadm_autoconfig() function will dynamically update
/var/etc/mdadm.conf, rather than using uci.  If
mdadm.@defaults[0].autoconfig is false then the previous manual
logic will be used.

START & STOP order were changed to allot more space
before fstab starts, i.e for a cryptsetup block layer, and
to stop mdadm last, to fix conflicts with mdadm
attempting to go offline prior to umount running.

An active, defunct device & uuid were left in the mdadm.config that
prevented mdadm from starting properly.  These were commented.

Signed-off-by: Joseph Tingiris <joseph.tingiris@gmail.com>

josephtingiris committed:
mdadm: corrected indentations

Signed-off-by: Joseph Tingiris <joseph.tingiris@gmail.com>

cshoredaniel commented:
Ah...don't forget to rebase your feature branch rather than keep adding commits: https://github.com/openwrt/packages/blob/master/CONTRIBUTING.md#advice-on-pull-requests

cshoredaniel commented on package/utils/mdadm/files/mdadm.init:
```diff
@@ -1,7 +1,7 @@
-START=13
-STOP=98
+START=12
+STOP=99
```
Perhaps umount should be moved up one ... the problem is that at 99 I'm not sure it's guaranteed to occur before the done script.

cshoredaniel replied:
Yeah, it was probably 6-10 years ago that I wrote up https://openwrt.org/docs/techref/requirements.boot.process, but was talking with a fellow by the name of Bernard Loos(sp?) about the need to deal with dependencies since that initial document was exposing some ordering challenges already.  Part of the issue for OpenWrt is having an init system that doesn't need an initramfs (for space/waste reasons), which is not a challenge systemd even attempts (ot my knowledge).  I'd have to dig to find those discussion / design documents, but we were thinking about this problem before systemd came around, but 'reel jobs' interfered with actually working on the problem.  Unfortunately I don't think I ever posted the dependency bits on the wiki, else procd might have supported them.

josephtingiris replied:
Thanks for sharing that.  It helped me understand where you're coming from.  I completely understand the challenges involved.  I've been programming embedded systems for decades, with a sharp focus on Linux.  I have a reel job too & know how it goes.

OpenWrt 18 already has most of what [systemd requires](https://github.com/systemd/systemd/blob/master/README).  Others have done it and [do it with busybox](https://github.com/profusion/demystifying-systemd-for-embedded-systems).  The nay saying is tiresome & the writing has been all over the walls for years.  Every modern distro uses it and for good reason.  

I'm glad I learned it.  I've grow to appreciate it and my finger memory is slowly forgetting _service <name> start_; that's my biggest complaint.  So many years of _service_ and with systemd, they flipped the arguments to _systemctl start <name>_ (which, imho, is better because there are less keystrokes to backspace to stop/start a service, etc.  Yes, it brings additional complexity, with the benefit of great new capabilities.  Maybe I'll whip up a systemd branch for Openwrt in my infinite spare time. ;)

At the moment, I want to get through this PR and continue learning how you all like things done.  So far, I'm impressed.  Back to the task at hand; incremental change.  Good, better, best.  This is the result of my init discovery with a fairly broad proposal.   While systemd may be best, this is certainly better.

```
v18.06.x & master (proposal)
============================
network                        START=12 STOP=93 * can network be up this early & stay up a little later?
blockd.init                    START=41 STOP=94 * must start after fstab; stop before lvm2
kdump.init                     START=41 STOP=96 * must start after fstab; must stop before umount
umount                         START=NO STOP=96 * must stop before mdadm
lvm2.init                      START=15 STOP=95 * must start before fstab; must stop after umount, but before cryptsetup & mdadm
cryptsetup.init                START=14 STOP=96 * must start before lvm2; second; must stop after umount & lvm2, but before mdadm
**mdadm.init                     START=13 STOP=99 * must start before cryptsetup; must stop (last) after umount, cryptsetup, & lvm2**
done                           START=99         * done nice coceptual name, but should still run dead last; i.e. S99Zz

v18.06.1 & master (current)
===========================
cryptsetup.init                START=?? STOP=?? * Bug. cryptsetup.init is missing entirely
umount                         START=NO STOP=99 * Bug. other block devices must stop afterwards.
boot                           START=10 STOP=98 * OK? point of reference; runs /sbin/mount_root; has STOP value set, though unused?
sysctl                         START=11         * OK. point of reference
mdadm.init                     START=13 STOP=98 * Bug. umount is K99
lvm2.init                      START=15         * Bug. STOP undefined
network                        START=20 STOP=90 * starts after dnsmasq, dropbear, firewall, etc?  nice to have up before cryptsetup, so key files can be curl'ed?                                                                                                             
fstab.init                     START=40         * OK. point of reference; runs /sbin/block mount (and /sbin/block umount)
kdump.init                     START=41 STOP=98 * must be before umount & change if umount does
blockd.init                    START=80         * automounter.  should start much sooner, after fstab, & explicitly stop.
done                           START=95         * runs /etc/rc.local  S95?  many things are run after S95 ... START rc.local rant
led                            START=96         * rc.local rant
rssileds.init                  START=96 STOP=96 * rc.local rant
dsl_control                    START=97         * rc.local rant
linksys_recovery               START=97         * rc.local rant
hwmon_fancontrol               START=98         * rc.local rant
sysntpd                        START=98         * rc.local rant
adb-enablemodem                START=99         * rc.local rant; only this would run before S99boot
bootcount                      START=99         * rc.local rant
bugcheck.initd                 START=99         * rc.local rant
igmpproxy.init                 START=99         * rc.local rant
netdev-cpu                     START=99         * rc.local rant
omcproxy.init                  START=99         * rc.local rant
set-irq-affinity               START=99         * rc.local rant
urandom_seed                   START=99         * OK. saves /etc/urandom.seed; entropy.
```

cshoredaniel commented on package/utils/mdadm/files/mdadm.init:
```diff
@@ -71,14 +99,14 @@ mdadm_array() {
+		config_foreach mdadm_array array
+	fi
 
 	$PROG --assemble --scan --config="$CONF"
```
Can you make CONF optional?

cshoredaniel replied:
will comment more later (but kudos for all the work you're doing on this).

Thist is what I'm referring to though '--assemble with --scan' is special-cased to do:

```
   Auto Assembly
       When  --assemble is used with --scan and no devices are listed, mdadm will first attempt to assemble
       all the arrays listed in the config file.

       If no arrays are listed in the config (other than those marked <ignore>) it will  look  through  the
       available devices for possible arrays and will try to assemble anything that it finds.  Arrays which
       are tagged as belonging to the given homehost will be assembled and started normally.  Arrays  which
       do not obviously belong to this host are given names that are expected not to conflict with anything
       local, and are started "read-auto" so that nothing is written to any device until the array is writ‐
       ten to. i.e.  automatic resync etc is delayed.
```

cshoredaniel replied:
Although I'm think that would be best with an empty config file, so skipping --conf bit isn't really necessary (I think back when I did it, I relied on shipping an entirely commented (or non-existent) mdadm.conf file so that the kernel would auto assemble.  However that was based on a fairly linear setup and doesn't deal with e.g. hotplug which is the norm in openwrt (in my case I was using old x86 hardware with internal hard drives).

josephtingiris replied:
I think _--config=none_ would produce the desired results with minimal effort.  It would be fairly trivial to tweak `$CONF=none` rather than rely on _/var/etc/mdadm.conf_.  It's certainly worth testing.

Introducing a second uci defaults option increases complexity and flexibility, e.g. `option config none` could pass through the actual (valid) values of **none**, **partition**, or (writable) **file**.  If `config` is unset or empty then stay using the global value of _/var/etc/mdadm.conf_.  If `autoconfig` is true (or false) **and** `config` is 'none' then attempt stateless Auto Assembly regardless? Or, should it always prefer to use a stateful config if `autoconfig` is true and `config` is 'none'?

Another option would be to use just one `config` option with 3 (or more) validated values and some would pass through.  Valid values would be **auto**, **partition**, **none**, or a (writable) **file** location.

cshoredaniel replied:
I'm not a core dev (was at one point, but 'resigned' as I was not well.  Thing are much better, but I haven't tried to get back in...also not really doing much with the core system these days), so take what I say with a gain of salt, but my preference is the latter.  I think the use is much more clear. @jow @blogic thoughts?

diizzyy commented:
@josephcsible 
Additionally (if you want) since you're already playing around with mdadm perhaps bump it to 4.1?

josephcsible commented:
@diizzyy I don't think I'm who you meant to ping.

diizzyy commented:
@josephcsible 
Correct, sorry about that :/

josephtingiris reviewed (commented):
Yeah, I will rev it up.  I'll refocus on that when I'm done sorting out the device chain inits.  kmod-md-raid456 as well?

cshoredaniel commented:
On 2019-01-04 11:42 p.m., Joseph Tingiris wrote:
>
> *@josephtingiris* commented on this pull request.
>
> ------------------------------------------------------------------------
> |v18.06.x & master (proposal) |
> |============================ network START=12 STOP=93 * can network
> be up this early & stay up a little later? blockd.init START=41
> STOP=94 * must start after fstab; stop before lvm2 kdump.init START=41
> STOP=96 * must start after fstab; must stop before umount umount
> START=NO STOP=96 * must stop before mdadm lvm2.init START=15 STOP=95 *
> must start before fstab; must stop after umount, but before cryptsetup
> & mdadm cryptsetup.init START=14 STOP=96 * must start before lvm2;
> second; must stop after umount & lvm2, but before mdadm **mdadm.init
> START=13 STOP=99 * must start before cryptsetup; must stop (last)
> after umount, cryptsetup, & lvm2** done START=99 * done nice coceptual
> name, but should still run dead last; i.e. S99Zz v18.06.1 & master
> (current) =========================== cryptsetup.init START=?? STOP=??
> * Bug. cryptsetup.init is missing entirely umount START=NO STOP=99 *
> Bug. other block devices must stop afterwards. boot START=10 STOP=98 *
> OK? point of reference; runs /sbin/mount_root; has STOP value set,
> though unused? sysctl START=11 * OK. point of reference mdadm.init
> START=13 STOP=98 * Bug. umount is K99 lvm2.init START=15 * Bug. STOP
> undefined network START=20 STOP=90 * starts after dnsmasq, dropbear,
> firewall, etc? nice to have up before cryptsetup, so key files can be
> curl'ed? fstab.init START=40 * OK. point of reference; runs
> /sbin/block mount (and /sbin/block umount) kdump.init START=41 STOP=98
> * must be before umount & change if umount does blockd.init START=80 *
> automounter. should start much sooner, after fstab, & explicitly stop.
> done START=95 * runs /etc/rc.local S95? many things are run after S95
> ... START rc.local rant led START=96 * rc.local rant rssileds.init
> START=96 STOP=96 * rc.local rant dsl_control START=97 * rc.local rant
> linksys_recovery START=97 * rc.local rant hwmon_fancontrol START=98 *
> rc.local rant sysntpd START=98 * rc.local rant adb-enablemodem
> START=99 * rc.local rant; only this would run before S99boot bootcount
> START=99 * rc.local rant bugcheck.initd START=99 * rc.local rant
> igmpproxy.init START=99 * rc.local rant netdev-cpu START=99 * rc.local
> rant omcproxy.init START=99 * rc.local rant set-irq-affinity START=99
> * rc.local rant urandom_seed START=99 * OK. saves /etc/urandom.seed;
> entropy. |
Would you be willing to send this or something based on it to the
openwrt-devel@lists.openwrt.org list?  That's the best place to discuss
distro wide stuff like this. 
https://lists.openwrt.org/mailman/listinfo/openwrt-devel

I'm glad someone is taking this up!

Regards,

Daniel

diizzyy commented:
@cshoredaniel 
Unless someone says so it's not a requirement at all and even if there is people need to adopt both paths as it's agreed to use both. Doing otherwise is just counterproductive and wastes peoples time not to mention contradicts suggested paths for contributing.

diizzyy commented:
@cshoredaniel 
https://openwrt.org/submitting-patches
That's the official guidelines

cshoredaniel commented:
@diizzyy However the point is question is not a single patch but potentially a number of them -- the discussion extends beyond the current PR or any single PR, hence my pointing to the list.

josephtingiris commented:
I'd originally created this PR from my fork's master branch.  After reading the official guidelines (sorry), I followed the directions and created a new branch: https://github.com/josephtingiris/openwrt/tree/PR1713

I had some time today and finished what I was working on there.  I've incorporated everything discussed here, tested, documented, and signed off.  Hopefully it's not too much for one day.

In any event, we could ...

1) Merge that back to my master and it should update this PR automatically.  Or, 
2) Submit a new PR.  Or, 
3) An admin could update this PR and change the base to my PR1713 branch.  

Which is preferred?

cshoredaniel commented:
@jow- @chunkeey @dedeckeh (most likely to respond quickly I think)

I've only skimmed the new commit so far but I'm impressed.  Excellent work!  (Nice finding X99 too).

cshoredaniel commented:
@josephtingiris The lack of comments is because there is a time when OpenWrt ran on devices with 2 MB of flash and 8 MB of RAM.   Currently 8 / 64 is a major part of the userbase and there are folks not happy that 4 / 32 is no longer possible without lots of custom work.  If there were more dev time I think it'd be useful support multiple paths (micro, mini, normal, and big, for example), but reality is that there probably isn't enough dev resources for that.

josephtingiris commented:
Thanks.  I agree.  Most products get by with S,M,L.  You know it's time to review the roadmap when the need to support XS,XL, & XXL.  The new AC2600+ stuff is awesome.  On the other end it's IoT gear.  Linux has a place across the board.  It all depends on how well the SDLC works and if everyone gets along.  What happened to Vertical Communications?

There are only 3 files & 1 of them is the Makefile PKG_RELEASE bump.  These are the significant two on my PR1713 branch.

https://github.com/josephtingiris/openwrt/blob/PR1713/package/utils/mdadm/files/mdadm.config
https://github.com/josephtingiris/openwrt/blob/PR1713/package/utils/mdadm/files/mdadm.init

cshoredaniel commented:
@diizzyy http://lists.infradead.org/pipermail/openwrt-devel/2019-January/015437.html -- basically the reply from Jo is that most devs will miss conversations in GitHub PR's.

neheb commented:
Just saw this thread.

I've given up on pull requests here. The last two came out of laziness (used GitHub's edit function).

That being said, patchwork still has dormant patches.

josephtingiris commented:
Yeah, it's a bit confusing.  I went ahead and submitted this PR to the list and see it in patchwork.  Still wondering if I should merge back to my master & update this PR or close it?

neheb commented:
Probably close. This won’t get any attention here.

Cabalist commented:
That patch in patchwork is here: https://patchwork.ozlabs.org/patch/1022738/ in case any one finds this issue like I did.

Currently sitting with _Changes Requested_ 

@josephtingiris I'd really like to see this merged.  It is very helpful for me with a Gnubee.  👍 Thanks for your work on this.

josephtingiris commented:
On 1/13, I resubmitted all changes requested as https://patchwork.ozlabs.org/project/openwrt/list/?series=85884

It’s been sitting in a New state;  I’ve been using it with no issues (and with the bumped version of mdadm).


________________________________
From: Ryan Jarvis <notifications@github.com>
Sent: Tuesday, February 26, 2019 4:59 PM
To: openwrt/openwrt
Cc: Joseph Tingiris; Mention
Subject: Re: [openwrt/openwrt] mdadm: add new init features; bug fixes (#1713)


That patch in patchwork is here: https://patchwork.ozlabs.org/patch/1022738/ in case any one finds this issue like I did.

Currently sitting with Changes Requested

@josephtingiris<https://github.com/josephtingiris> I'd really like to see this merged. It is very helpful for me with a Gnubee. 👍 Thanks for your work on this.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/openwrt/openwrt/pull/1713#issuecomment-467631300>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABxWmPIr25KVPRZWNi32NZcHVbfCymuAks5vRa5JgaJpZM4Zof-v>.

Cabalist commented:
Yow.  Patchwork is confusing coming from Gitlab/Github   I would never have found that.  Thanks!  

What would be the best way to let the devs know we'd like this merged?

neheb commented:
ping @hauke maybe.

@jow- had comments on the mdadm patchset but he's away right now.

josephtingiris commented:
I've resubmitted
https://patchwork.ozlabs.org/project/openwrt/list/?series=94368 that
incorporated what I could of @jow's previous comments and what I've been
using with the new version of mdadm.  This would be mdadm 4.1 PKG_RELEASE 2.

@hauke could you please review or point me in the right direction?

Thanks.

On Tue, Feb 26, 2019 at 8:22 PM Rosen Penev <notifications@github.com>
wrote:

> ping @hauke <https://github.com/hauke> maybe.
>
> @jow- <https://github.com/jow-> had comments on the mdadm patchset but
> he's away right now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/1713#issuecomment-467684490>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABxWmMbmqJJAj42jgfJ1DRwjllcghDZEks5vRd3RgaJpZM4Zof-v>
> .
>
